### PR TITLE
[#23] [Chore] Set up Docker compose for local development

### LINKED ISSUE #23: [Chore] Set up Docker compose for local development
## Why
Setting up a Docker Compose file will help other developers start the project easily and start the development. 

## Acceptance Criteria

- Set up a Docker Compose file for local development: `docker-compose.dev.yml`
- Set up service for the database.
- Set up service for the worker.
- (Optional) Set up a service for the application. 
- Add a Makefile to start the containers (see https://makefiletutorial.com/). The command could be: `make start`

### PR DESCRIPTION
**Pull request by Prabeshpd**
- Close #23 

## What happened 👀

- Added a config for redis connection
- Used redis connection on worker
- Updated workflow for redis
- Added docker file for application
- Added docker compose for development
- Added makefile

## Insight 📝

- Redis connection is passed for bullmq workers
     - SInce, we are using docker. Application will not have access to docker containers through localhost
     - Bullmq directly looked for localhost connection
     - For this we are adding a redis connection
     - For redis connection singleton pattern is used.
 - Added a docker-compose file
- Added a make file
> [!Note]
> - I have not used the docker image of application on development. Since, we want the latest changes of development to visible. 
> - We can mount the volume of directory source to docker source directory. I have avoided that.
> - I have used the local server and docker instance of redis and postgres.

## Proof Of Work 📹

https://github.com/nimblehq/ic-nextjs-prabesh-dahal/assets/39412015/b6762ed3-6fe4-402f-a95a-fbd862018d15

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,56 @@
+FROM node:18-bullseye-slim AS base
```
Comment (Prabeshpd) [inline Dockerfile:1]: For the spacing :
I did not find any specifications on compass for dockerfile

- I have added a line break to separate the docker image layers.
- I have tried to group the same command layer and added extra line to separate them.
- As, I have built a multi step docker I have added two lines between each step.

File: tailwind.config.js [hunk 2]
```diff
@@ -5,14 +5,14 @@ module.exports = {
     './src/components/**/*.{js,ts,jsx,tsx,mdx}',
     './src/app/**/*.{js,ts,jsx,tsx,mdx}',
   ],
+  darkMode: 'class',
```
Comment (olivierobert) [inline tailwind.config.js:8]: This change is outside the scope of change for this PR 🤔 It is best to not buncle this in the same set of changes (I also do know why this is needed or what this will do).

File: .env.example [hunk 3]
```diff
@@ -8,3 +8,10 @@ NEXT_AUTH_SECRET=
 
 # Request  in Next.JS > 13 does not support the query attribute, hence requires a fully qualified URL
 TEST_APP_HOST=
+
+# Docker env variables
```
Comment (olivierobert) [inline .env.example:12]: It is confusing to add this description as passing env variables to a Docker container has its own syntax (see https://vsupalov.com/docker-arg-env-variable-guide/#setting-env-values)

In addition, I do not see these env variables being used in the codebase 🤔 So why are these PG related env needed?

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+# Install required dependecies
+FROM base as dependencies
+
+#Puppeteer installation
```
Comment (olivierobert) [inline Dockerfile:6]: This comment is rather misleading as the code skips the installation of Puppeteer (but the tile says otherwise)

```suggestion
# Disable Puppeteer installation
```

Also, why disabling it? Wouldn't it be required for the PDF export?

File: Dockerfile [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    apt-get install chromium -y --no-install-recommends && \
+    rm -rf /var/lib/apt/lists/*
+
+WORKDIR /app
+
+COPY package.json yarn.lock ./
+
+RUN yarn --frozen-lockfile
+
+
```
Comment (olivierobert) [inline Dockerfile:23]: No convention on that front, but the extra line break can be removed. The comment block is sufficient to delimit multi-stage build

File: Dockerfile
Comment (olivierobert) [inline Dockerfile:1]: Why is this file added in this PR? Shouldn;t this be for #126? Since this PR is for local development and the app is not run in a container, there is no need for such a file yet.

File: Dockerfile [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+ENV NODE_ENV production
+
+WORKDIR /app
+
+RUN addgroup --system --gid 1001 nodejs && \
+    adduser --system --uid 1001 nextjs
+RUN mkdir .next && \
+    chown nextjs:nodejs .next
+
+COPY --from=builder --chown=nextjs:nodejs /app .
```
Comment (olivierobert) [inline Dockerfile:48]: Wouldn't that copy also all subdirectories in `build`?

File: Makefile [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+
+.PHONY: dev-start dev-stop test
+
+dev-start:
+	docker compose -f $(dev_compose_file) up -d
+	yarn migrate:dev
+	yarn dev
+
+dev-stop:
+	docker compose -f $(dev_compose_file) down -v
```
Comment (olivierobert) [inline Makefile:13]: For local dev, there would be no need to remove each time, so the following would suffice:

```suggestion
	docker compose -f $(dev_compose_file) stop -v
```

File: src/config/redis.ts [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+  host: process.env.REDIS_HOST || 'localhost',
+  port: Number(process.env.REDIS_PORT || 6379),
+  password: process.env.REDIS_PASSWORD || '',
+};
+
+export default class Redis {
+  private static instance: IORedis;
+
+  private constructor() {
+    throw new NotImplementedError({ message: 'Class can not be instantiated' });
```
Comment (olivierobert) [inline src/config/redis.ts:21]: Having a class that cannot be instantiated is unexpected. Two options: 1) move the  code for `getInstance` to the constructor method or 2) remove the constructor method altogether. 

Here is below, an example from the easyHotel APi codebase:

File: src/workers/queueScraper.worker.ts [hunk 9]
```diff
@@ -1,15 +1,23 @@
 import { Worker, Queue } from 'bullmq';
 
+import Redis from '@/config/redis';
 import { AsyncProcess } from '@/constants/enum';
 import { BullQueue } from '@/lib/worker';
 
 import buildScraper from './queueScraper';
 
-const scrapeBullQueue = new Queue(AsyncProcess.QUEUE_SCRAPE);
+const redis = Redis.getInstance();
```
Comment (olivierobert) [inline src/workers/queueScraper.worker.ts:9]: Let's be more descriptive, especially as there is another variable close to this one (`Redis`):

```suggestion
const redisConnection = Redis.getInstance();
```

File: .env.example (hunk 3 above)
Comment (Prabeshpd) [reply .env.example:12]: It's needed for the docker-compose file to add environment. Since, there are few environment we can directly inject to compose file. Makes sense will remove it.

File: Dockerfile (hunk 6 above)
Comment (Prabeshpd) [reply Dockerfile:48]: Yes, but next application build has lots of internal folders which if we do not copy will not render the CSS images.
Had a lot of copy statements lines so, ended up copying all the folders.

File: Dockerfile (hunk 4 above)
Comment (Prabeshpd) [reply Dockerfile:6]: It installs chromium under the hood. The environment variable for PUPPETEER_SKIP_CHROMIUM_DOWNLOAD is to skip the latest version of chromium. But I guess it can be removed.

File: src/config/redis.ts (hunk 8 above)
Comment (Prabeshpd) [reply src/config/redis.ts]: I can not see the image. 😅 . Have removed the constructor for now.

**Comment by Prabeshpd** [comment]
Have moved the docker file to https://github.com/nimblehq/ic-nextjs-prabesh-dahal/pull/143

File: src/workers/queueScraper.worker.ts (hunk 9 above)
Comment (Prabeshpd) [reply src/workers/queueScraper.worker.ts]: Fixed.

File: Dockerfile (hunk 5 above)
Comment (Prabeshpd) [reply Dockerfile]: Got it.

File: tailwind.config.js (hunk 2 above)
Comment (Prabeshpd) [reply tailwind.config.js]: Removed it.

File: Makefile (hunk 7 above)
Comment (Prabeshpd) [reply Makefile]: Fixed

File: tailwind.config.js [hunk 10]
```diff
@@ -9,10 +9,9 @@ module.exports = {
     extend: {
       backgroundImage: {
         'gradient-radial': 'radial-gradient(var(--tw-gradient-stops))',
-        'gradient-conic':
-          'conic-gradient(from 180deg at 50% 50%, var(--tw-gradient-stops))',
+        'gradient-conic': 'conic-gradient(from 180deg at 50% 50%, var(--tw-gradient-stops))',
```
Comment (Prabeshpd) [inline tailwind.config.js:12]: Code format revamped

File: src/config/redis.ts (hunk 8 above)
Comment (olivierobert) [reply src/config/redis.ts]: I have reuploaded the image

File: src/config/redis.test.ts [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+import Redis from './redis';
+
+describe('Redis:', () => {
+  let redis: IORedis;
+
+  afterAll(() => {
+    redis.disconnect();
+  });
+
+  describe('given "getInstance" is called multiple times', () => {
```
Comment (tednguyendev) [inline src/config/redis.test.ts:15]: Should we also test the case that: When there is no instance, it init a `IORedis` 👀

File: README.md [hunk 12]
```diff
@@ -38,6 +38,18 @@ The application is bootstrapped with [`create-next-app`](https://github.com/verc
...
    ```
 
+## Docker Compose
+
+1. Start Development Server
+
+   ```bash
+   cp .env.example .env
+
+   make dev-start
```
[the remark this answers is not in the document]
Comment (Prabeshpd) [reply README.md:48]: You may need to add POSTGRES_DB_CONNECTION_URL in .env.test. That usually depends on your configuration. So, have added empty string in .env.example.

I will add a extra information to make sure to add the URL in Readme.

File: Makefile [hunk 13]
```diff
@@ -0,0 +1,12 @@
+dev_compose_file := docker-compose.dev.yml
+application_container_test := application_test
```
[the remark this answers is not in the document]
Comment (Prabeshpd) [reply Makefile]: Removed it.

File: README.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (Prabeshpd) [reply README.md]: Updated the Readme

File: src/config/redis.test.ts (hunk 11 above)
Comment (Prabeshpd) [reply src/config/redis.test.ts:23]: Oh. you mean to ensure that the instance is of IORedis. Makes sense. Have updated it.

File: README.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (tednguyendev) [reply README.md]: Yes it works now.

1 point is that I think we should move [this part](https://github.com/nimblehq/ic-nextjs-prabesh-dahal/pull/141/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R9-R13) to under [this part](https://github.com/nimblehq/ic-nextjs-prabesh-dahal/pull/141/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R15-R19).

As in the current order: If we add the `POSTGRES_DB_CONNECTION_URL` to `.env`, then `cp .env.example .env`, the `.env` will be override by the `.env.example`, make it not has the `POSTGRES_DB_CONNECTION_URL` anymore 👍

File: README.md [hunk 14]
```diff
@@ -2,6 +2,24 @@
 
 The application is bootstrapped with [`create-next-app`](https://github.com/vercel/next.js/tree/canary/packages/create-next-app).
 
+## Docker Compose
+
+- ### Start Development Server
```
Comment (tednguyendev) [inline README.md:7]: How about just

```suggestion
## Docker

```

**Review by tednguyendev — APPROVED** [review]

File: README.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (Prabeshpd) [reply README.md]: Done.

File: README.md (hunk 14 above)
Comment (Prabeshpd) [reply README.md:7]: I have added readme for test part as well. So, it will help us to differentiate between local development and testing.

File: .env.example [hunk 15]
```diff
@@ -8,3 +8,7 @@ NEXT_AUTH_SECRET=
 
 # Request  in Next.JS > 13 does not support the query attribute, hence requires a fully qualified URL
 TEST_APP_HOST=
+
+REDIS_HOST=
```
Comment (olivierobert) [inline .env.example:12]: It would be better to restructure the env as now it seems that Redis is part of the test env variable:

```
POSTGRES_DB_CONNECTION_URL=
REDIS_HOST=
REDIS_PORT=
REDIS_PASSWORD=

REACT_BASE_ENV_URL=
NEXT_AUTH_SECRET=

OAUTH_GOOGLE_CLIENT_ID=
OAUTH_GOOGLE_CLIENT_SECRET=

# Request  in Next.JS > 13 does not support the query attribute, hence requires a fully qualified URL
TEST_APP_HOST=
```

Also, could `TEST_APP_HOST=` be moved to an `.env.test`? Not a blocker but it could be a good move.

**Review by olivierobert — APPROVED** [review]
Just a minor suggestion for you to consider. Up to you to decide to add it in or not.

File: .env.example (hunk 15 above)
Comment (Prabeshpd) [reply .env.example]: Updated it.
